VestingInstance.sol

```diff
@@ -0,0 +1,273 @@
+    }
+
+    /**
+     * @notice Allows only the revoker to revoke the vesting. Gold already vested
+     * remains in the contract, the rest is returned to the _refundDestination.
+     * @param revokeTime the revocation timestamp
+     * @dev revokeTime the revocation timestamp. If is less than the current block timestamp, it is set equal
+     */
+    function revoke(revokeTime) external onlyRevoker {
+        require(_revocable, "Revoking is not allowed");
+        require(!_revoked, "Vesting already revoked");
+
+        uint256 revokeTimestamp = revokeTime > block.timestamp ? revokeTime : block.timestamp;
+
+        uint256 balance = getGoldToken().balanceOf(address(this));
+        uint256 releasableAmount = _getReleasableAmount(revokeTimestamp);
+        uint256 refund = balance.sub(releasableAmount);
+
+        _revoked = true;
+        _revokeTime = revokeTimestamp;
+
+        getGoldToken().transfer(_refundDestination, refund);
+
+        emit VestingRevoked(msg.sender, _refundDestination, refund, _revokeTime);
+    }
+
+    /**
+     * @dev Calculates the amount that has already vested but hasn't been withdrawn (released) yet.
+     * @param timestamp the timestamp at which the calculate the releasable amount
+     */
+    function _getReleasableAmount(uint256 timestamp) private view returns (uint256) {
+        return _calculateFreeAmount(timestamp).sub(_currentlyReleased);
+    }
+
+    /**
+     * @dev Calculates the amount that has already vested.
+     * @param timestamp the timestamp at which the calculate the already vested amount
+     */
+    function _calculateFreeAmount(uint256 timestamp) private view returns (uint256) {
+        uint256 currentBalance = getGoldToken().balanceOf(address(this));
+        uint256 totalBalance = currentBalance.add(_currentlyReleased);
+
+        if (timestamp < _cliffStartTime) {
+            return 0;
+        } else if (timestamp >= _vestingStartTime.add( _vestingPeriods.mul(_vestingPeriodSec) ) || _revoked) {
+            return totalBalance;
+        } else {
+            uint256 gradient = (timestamp.sub(_vestingStartTime)).div(_vestingPeriodSec);
+            return  ( (currentBalance.mul(gradient)).mul(vestAmountPerPeriod) ).div(totalBalance);
+        }
+    }
+
+    /**
+     * @notice A wrapper func for the lock gold method
+     * @param value the value to gold to be locked
+     * @return True if the transaction succeeds.
+     * @dev To be called only by the beneficiary of the vesting
+     */
+    function lockGold(uint256 value) external onlyBeneficiary returns (bool) {
+
+      // the beneficiary may not lock more than the vesting has currently released
+      uint256 unreleasedAmount = _getReleasableAmount(block.timestamp);
+      require(unreleasedAmount >= value, "Gold Amount to lock must not be less that the currently releasable amount");
+
+      bool success;
+      (success,) = address(getLockedGold()).lock.gas(gasleft()).value(msg.value)();
```

Comment:
Should we emit VestingGoldLocked if success = false? I guess `require(success)` might prove useful

Reply:
sure, added require to all of them